### ISSUE TITLE
@inlineCallbacks ignores return values in Python3.3 and greater

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7624]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#7624|
|Type|enhancement|
|Created|2014-09-02 20:01:59Z|
|Branch|https://github.com/twisted/twisted/tree/inlineCallbacks-return-7624|

In Python3.3+ the `return` statement of generators can include a value, but `@inlineCallbacks` ignores such values.

This has two results: a) silently ignoring stuff is bad and confusing, b) it means folks have to use `returnValue` which has more complex semantics than simple return values.

Attachments:

* [t7624.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c6/5c696fdafa4779e293b52277671f30693b568cfc/7c72378cc4b530de0d0b1d9315ffebc217167c2f.diff) (2197 bytes) - added by Alex on 2014-09-02 20:02:25Z - 
* [t7624.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c6/5c696fdafa4779e293b52277671f30693b568cfc/8cf2273ad54d867a6df223d5bffc3011d1a5baae.diff) (4080 bytes) - added by Alex on 2014-09-08 18:16:21Z - 
* [t7624.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/5c6/5c696fdafa4779e293b52277671f30693b568cfc/b48191ee187242d53bce16f6feae0f6e205e7b1f.diff) (4110 bytes) - added by Alex on 2014-09-08 19:40:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7624 7624
type__enhancement enhancement
reporter__Alex Alex
priority__normal normal
milestone__ 
branch__branches_inlineCallbacks_return_7624 branches/inlineCallbacks-return-7624
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1409688119840553 1409688119840553
changetime__1474198049187875 1474198049187875
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to Alex|
|-|-|

Thanks.  This is missing prose/example-style documentation and a news fragment.  It also seems the information in the class docstring would be more useful in the docstring of `inlineCallbacks` itself.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

In other words, this is blocked on …

…

[#7343](https://github.com/twisted/twisted/issues/7343)

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Alex set owner to @glyph|
|-|-|

Added docs and a topfile

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to Alex|
|-|-|

Thanks for addressing the concerns in the previous review.

This is almost OK to land; but I am slightly concerned about the fact that a python 3.4+ user might discover this documentation, put a `return` statement in an `inlineCallbacks` function, and then get wrong results, since they have a Twisted older than 14.1.  I don't think this qualifies as "compatibility breakage" necessarily because you'd have to really go out of your way to depend absolutely on a `None` return value from a function trying to return a value with a return statement.  Do you think you could add a brief note which explains the timing of the change, though?

Resubmit for review with said change and I can probably land.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|Alex set owner to @glyph|
|-|-|

Added a `versionadded` indicator.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

(In [43017]) Branching to inlineCallbacks-return-7624.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [43019]) Merge inlineCallbacks-return-7624: allow 'return' in python 3.3+

Author: Alex

Reviewer: glyph

Fixes: [#7624](https://github.com/twisted/twisted/issues/7624)

Add the ability to use 'return' to supply a Deferred result, obviating
the need for returnValue, in versions of Python which syntactically
support this.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing review from closed tickets.
